File: leet-code/56.合并区间.js
/*
 * @lc app=leetcode.cn id=56 lang=javascript
 *
 * [56] 合并区间
 */

// @lc code=start
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function (intervals) {
  intervals.sort((a, b) => {
    if (a[0] === b[0]) {
      return a[1] - b[1]
    } else {
      return a[0] - b[0]
    }
  })
  let i = 0
  const res = []
  while (i < intervals.length) {
    const interval1 = intervals[i]
    let right = interval1[1]
    let j = i + 1
    while (j < intervals.length) {
      const interval2 = intervals[j]
      if (interval2[0] > right) break
      j++
      // 更新右边界
      right = Math.max(right, interval2[1])
    }
    res.push([interval1[0], right])
    i = j
  }
  return res
};
// @lc code=end

File: leet-code-rs/src/solutions/56_合并区间.rs
/*
 * @lc app=leetcode.cn id=56 lang=rust
 *
 * [56] 合并区间
 */

// @lc code=start

#[allow(unused)]
impl Solution {
    pub fn merge(mut intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        intervals.sort_by(|a, b| a[0].cmp(&b[0]));

        let mut res = vec![];
        for i in 0..intervals.len() {
            if i == 0 {
                res.push(intervals[i].clone());
                continue;
            }

            let mut pre = res.last_mut().unwrap();
            if intervals[i][0] <= pre[1] {
                // 有交集
                pre[1] = pre[1].max(intervals[i][1]);
            } else {
                res.push(intervals[i].clone());
            }
        }

        return res;
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_merge() {
        //assert_eq!(Solution::merge(vec![]),[]);
        assert!(true)
    }
}
